### ISSUE TITLE
Clarify and refine “temporary” handling in ref_as_ptr lint

### ISSUE THREAD

**Issue by Abd002** [rust-lang/rust-clippy#16220]
### Description

 The `ref_as_ptr` lint currently warns on *all* reference → pointer casts, including borrows of named locals. There is an internal TODO in the implementation:

 ```
 // TODO: only block the lint if `cast_expr` is a temporary
 ```

 I interpret “temporary” to mean *values created by expressions without a named local*, e.g., literals (`&42`), function return values (`&foo()`), struct/tuple literals, etc., whose lifetime ends at statement evaluation — not references to named variables. Under this definition:

 ```rust
 let val = 1;
 f(&val as *const _);       // should *not* lint
 f(&val as *const i32);     // should *not* lint
 ```

 but

 ```rust
 f(&42 as *const _);        // should lint
 let _ = &String::new() as *const _;  // should lint
 ```

 because those expressions borrow temporaries.

 The current implementation lints both cases. If the goal of the TODO is to only warn when `cast_expr` is a true temporary, I want to confirm if this understanding is correct.

### Version

```text

```

**Comment by Jarcho** [comment]
> let val = 1;
> f(&val as *const _);       // should *not* lint
> f(&val as *const i32);     // should *not* lint

These should both lint.

> f(&42 as *const _);        // should lint
> let _ = &String::new() as *const _;  // should lint

Only the first should lint here. The second is relying on temporary lifetime extension.

**Comment by Abd002** [comment]
The current implementation checks only the use context let/const/static to suppress the lint, but doesn't check whether the reference being cast is a temporary. For example:

 ```rust
 fn main() {
     let value = 42;
     let ptr = (&value) as *const _; // `&value` is a named local
 }
 ```

 Here `&value` is not a temporary, but because it appears in a `let` binding, the current logic will suppress the lint. My implementation adds a check for `is_expr_temporary_value`, so only borrows of *true temporaries* in lifetime-extending contexts are suppressed. This matches the TODO comment *“only block the lint if `cast_expr` is a temporary”* more accurately.

but there is another problem raised, lint suggest 
``` Rust
LL |     let ptr = (&val) as *const _;
   |               ^^^^^^^^^^^^^^^^^^ help: try: `std::ptr::from_ref((&val))`
```
the double parenthesis is invalid
